lib/ansible/modules/crypto/entrust/ecs_certificate.py

```diff
@@ -0,0 +1,906 @@
+        try:
+            self.ecs_client.GetAppVersion()
+        except RestOperationException as e:
+            module.fail_json(msg='Please verify credential information. Received exception when testing ECS connection: {0}'.format(to_native(e.message)))
+
+    # Conversion of the fields that go into the 'tracking' parameter of the request object
+    def convert_tracking_params(self, module):
+        body = {}
+        tracking = {}
+        if module.params.get('requester_name'):
+            tracking['requesterName'] = module.params['requester_name']
+        if module.params.get('requester_email'):
+            tracking['requesterEmail'] = module.params['requester_email']
+        if module.params.get('requester_phone'):
+            tracking['requesterPhone'] = module.params['requester_phone']
+        if module.params.get('tracking_info'):
+            tracking['trackingInfo'] = module.params['tracking_info']
+        if(module.params.get('custom_fields')):
+            # Omit custom fields from submitted dict if not present, instead of submitting them with value of 'null'
+            # The ECS API does technically accept null without error, but it complicates debugging user escalations and is unnecessary bandwidth.
+            custom_fields = {}
+            for k, v in module.params['custom_fields'].items():
+                if v is not None:
+                    custom_fields[k] = v
+            tracking['customFields'] = custom_fields
+        if(module.params.get('additional_emails')):
+            tracking['additionalEmails'] = module.params['additional_emails']
+        body['tracking'] = tracking
+        return body
+
+    def convert_cert_subject_params(self, module):
+        body = {}
+        if module.params.get('subject_alt_name') and len(module.params['subject_alt_name']) > 0:
+            body['subjectAltName'] = module.params['subject_alt_name']
+        if module.params.get('org'):
+            body['org'] = module.params['org']
+        if module.params.get('ou') and len(module.params['ou']) > 0:
+            body['ou'] = module.params['ou']
+        return body
+
+    def convert_general_params(self, module):
+        body = {}
+        if module.params.get('eku'):
+            body['eku'] = module.params['eku']
+        if self.request_type == 'NEW':
+            body['certType'] = module.params['cert_type']
+        body['clientId'] = module.params['client_id']
+        body.update(convert_module_param_to_json_bool(module, 'ctLog', 'ct_log'))
+        body.update(convert_module_param_to_json_bool(module, 'endUserKeyStorageAgreement', 'end_user_key_storage_agreement'))
+        return body
+
+    def convert_expiry_params(self, module):
+        body = {}
+        if module.params.get('cert_lifetime'):
+            body['certLifetime'] = module.params['cert_lifetime']
+        elif module.params.get('cert_expiry'):
+            body['certExpiryDate'] = module.params['cert_expiry']
```

Comment:
Maybe you want to validate `module.params['cert_expiry']` before handing it to the API; then user errors will be caught a lot earlier.

Reply:
I thought more about a regex which prevents the worst mistakes. But then, I don't know RFC 3339 well enough to be of help here :)